src/BlazorWebView/src/core/iOS/BlazorWebViewHandler.iOS.cs

```diff
@@ -153,10 +177,28 @@ public static void MapRootComponents(BlazorWebViewHandler handler, IBlazorWebVie
 			handler.StartWebViewCoreIfPossible();
 		}
 
-		public static void MapServices(BlazorWebViewHandler handler, IBlazorWebView webView)
+		private void OnRootComponentsCollectionChanged(object? sender, global::System.Collections.Specialized.NotifyCollectionChangedEventArgs eventArgs)
 		{
-			handler.Services = webView.Services;
-			handler.StartWebViewCoreIfPossible();
+			// If we haven't initialized yet, this is a no-op
+			if (_webviewManager != null)
+			{
+				// Dispatch because this is going to be async, and we want to catch any errors
+				_ = _webviewManager.Dispatcher.InvokeAsync(async () =>
+				{
+					var newItems = eventArgs.NewItems!.Cast<RootComponent>();
+					var oldItems = eventArgs.OldItems!.Cast<RootComponent>();
+
+					foreach (var item in newItems.Except(oldItems))
+					{
+						await item.AddToWebViewManagerAsync(_webviewManager);
+					}
+
+					foreach (var item in oldItems.Except(newItems))
+					{
+						await item.RemoveFromWebViewManagerAsync(_webviewManager);
+					}
+				});
+			}
```

Comment:
The amount of duplication across the three hosting flavours here makes me wonder if something common could be extracted from all three. Not suggesting you should hold up this PR for that, but might save time in the long run.

Reply:
I've cleaned it up and put the common code in a common file.